File: LogicAndTrick.WikiCodeParser/Elements/RefElement.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;
using LogicAndTrick.WikiCodeParser.Nodes;

namespace LogicAndTrick.WikiCodeParser.Elements
{
    public class RefElement : Element
    {
        public override bool Matches(Lines lines)
        {
            var value = lines.Value().Trim();
            return value.Length > 4 && value.StartsWith("[ref=") && Regex.IsMatch(value, @"\[ref=[a-z]+\]", RegexOptions.IgnoreCase);
        }

        public override INode Consume(Parser parser, ParseData data, Lines lines, string scope)
        {
            var current = lines.Current();
            var arr = new List<string>();

            var line = lines.Value().Trim();
            var res = Regex.Match(line, @"\[ref=([a-z ]+)\]", RegexOptions.IgnoreCase);
            if (!res.Success)
            {
                lines.SetCurrent(current);
                return null;
            }

            line = line.Substring(res.Value.Length);

            var name = res.Groups[1].Value;

            if (line.EndsWith("[/ref]"))
            {
                lines.Next();
                arr.Add(line.Substring(0, line.Length - 6));
            }
            else
            {
                if (line.Length > 0) arr.Add(line);
                var found = false;
                while (lines.Next())
                {
                    var value = lines.Value().TrimEnd();
                    if (value.EndsWith("[/ref]"))
                    {
                        var lastLine = value.Substring(0, value.Length - 6);
                        arr.Add(lastLine);
                        found = true;
                        break;
                    }
                    else
                    {
                        arr.Add(value);
                    }
                }

                if (!found || arr.Count == 0)
                {
                    lines.SetCurrent(current);
                    return null;
                }
            }

            // Store the ref node
            var node = parser.ParseElements(data, string.Join("\n", arr).Trim(), scope);
            data.Set($"Ref::{name}", node);

            // Return nothing
            return PlainTextNode.Empty;
        }
    }
}

File: javascript/src/Elements/RefElement.ts
import { Parser } from '..';
import { Lines } from '../Lines';
import { INode } from '../Nodes/INode';
import { PlainTextNode } from '../Nodes/PlainTextNode';
import { ParseData } from '../ParseData';
import { Element } from './Element';

export class RefElement extends Element {
    public Matches(lines: Lines): boolean {
        const value = lines.Value().trim();
        return value.length > 4 && value.startsWith('[ref=') && value.match(/\[ref=[a-z0-9 ]+\]/i) != null;
    }
    public Consume(parser: Parser, data: ParseData, lines: Lines, scope: string): INode {
        const current = lines.Current();
        const arr: string[] = [];

        let line = lines.Value().trim();
        const res = line.match(/\[ref=([a-z0-9 ]+)\]/i);
        if (!res) {
            lines.SetCurrent(current);
            return null;
        }

        line = line.substring(res[0].length);

        const name = res[1];

        if (line.endsWith('[/ref]')) {
            arr.push(line.substring(0, line.length - 6));
        } else {
            if (line.length > 0) arr.push(line);
            let found = false;
            while (lines.Next()) {
                const value = lines.Value().trimEnd();
                if (value.endsWith('[/ref]')) {
                    const lastLine = value.substring(0, value.length - 6);
                    arr.push(lastLine);
                    found = true;
                    break;
                }
                else {
                    arr.push(value);
                }
            }

            if (!found || arr.length == 0) {
                lines.SetCurrent(current);
                return null;
            }
        }

        // Store the ref node
        const node = parser.ParseElements(data, arr.join('\n').trim(), scope);
        data.Set(`Ref::${name}`, node);

        // Return nothing
        return PlainTextNode.Empty();
    }
}